crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs

```diff
@@ -262,126 +261,91 @@ struct EnumImpls {
-    let mut enum_index_of: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_name_at: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_field_len: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_variant_name: Vec<proc_macro2::TokenStream> = Vec::new();
-    let mut enum_variant_type: Vec<proc_macro2::TokenStream> = Vec::new();
+    let mut variant_info = Vec::new();
+    let mut enum_field = Vec::new();
+    let mut enum_field_at = Vec::new();
+    let mut enum_index_of = Vec::new();
+    let mut enum_name_at = Vec::new();
+    let mut enum_field_len = Vec::new();
+    let mut enum_variant_name = Vec::new();
+    let mut enum_variant_type = Vec::new();
 
     for variant in reflect_enum.active_variants() {
         let ident = &variant.data.ident;
         let name = ident.to_string();
         let unit = reflect_enum.get_unit(ident);
 
-        match &variant.fields {
-            EnumVariantFields::Unit => {
-                variant_info.push(quote! {
-                    #bevy_reflect_path::VariantInfo::Unit(
-                        #bevy_reflect_path::UnitVariantInfo::new(#name)
-                    )
-                });
-                enum_variant_name.push(quote! {
-                    #unit => #name
-                });
-                enum_variant_type.push(quote! {
-                    #unit => #bevy_reflect_path::VariantType::Unit
-                });
-            }
-            EnumVariantFields::Unnamed(fields) => {
+        // This is equivalent to a |fields: &Vec<StructField>, to_run: |usize, usize, &StructField| -> TokenStream|
+        // closure. Issue is that the closure cannot itself accept another closure, because closures cannot vary in
+        // the concrete type they accept, and each closure has a different concrete type.
+        macro_rules! for_fields {
+            ($fields:expr, |$reflect_idx:ident, $declar_field:pat, $field:ident| $body:expr) => {{
                 let mut field_info = Vec::new();
-                let mut field_idx: usize = 0;
-                for field in fields.iter() {
-                    if field.attrs.ignore {
+                let mut $reflect_idx: usize = 0;
+                for ($declar_field, $field) in $fields.iter().enumerate() {
+                    if $field.attrs.ignore {
                         // Ignored field
                         continue;
                     }
+                    field_info.push($body);
+                    $reflect_idx += 1;
+                }
+                ($reflect_idx, field_info)
+            }};
+        }
+        let (variant_type, field_len, field_info) = match &variant.fields {
+            EnumVariantFields::Unit => ("Unit", 0usize, quote!(#name)),
 
-                    let empties = utility::underscores(field_idx);
+            EnumVariantFields::Unnamed(fields) => {
+                let (field_len, field_info) = for_fields!(fields, |reflect_idx, declar, field| {
```

Comment:
Can we change `field_info` to `fields` or `field_constructors` or something here?

It's just slightly more readable than "`field_info` is used to create `field_info`".